bindings/cpp/newapi/session.cpp

```diff
@@ -844,4 +844,196 @@ async_iterator_result session::server_send(const std::vector<key> &keys, uint64_
+		return dnet_id_cmp(&lhs, &rhs) < 0;
+	}
+
+private:
+	std::vector<dnet_id> m_keys;
+	std::vector<bool> m_key_responses;
+	session m_session;
+	async_result_handler<read_result_entry> m_handler;
+	int m_last_error;
+	const dnet_addr m_addr;
+	std::unique_ptr<dnet_logger> m_log;
+};
+
+static async_read_result send_bulk_read(session &sess, const std::vector<dnet_id> &keys,
+					uint64_t read_flags) {
+	trace_scope scope{sess.get_trace_id(), sess.get_trace_bit()};
+
+	if (keys.empty()) {
+		async_read_result result{sess};
+		async_result_handler<read_result_entry> handler{result};
+		handler.complete(create_error(-ENXIO, "send_bulk_read: keys list is empty"));
+		return result;
+	}
+
+	auto dnet_addr_comparator = [] (const dnet_addr &lhs, const dnet_addr &rhs) -> bool {
+		return dnet_addr_cmp(&lhs, &rhs) < 0;
+	};
+	std::map<dnet_addr, std::vector<dnet_id>, decltype(dnet_addr_comparator)>
+		remotes_ids(dnet_addr_comparator); // node_address -> [list of keys]
+
+	async_read_result failed_result{sess.clean_clone()};
+	async_result_handler<read_result_entry> failed_handler{failed_result};
+
+	dnet_addr address;
+	int backend_id;
+	for (const dnet_id &id : keys) {
+		const int err = dnet_lookup_addr(sess.get_native(), nullptr, 0, &id, id.group_id, &address, &backend_id);
+		if (!err) {
+			remotes_ids[address].emplace_back(id);
+		} else {
+			memset(&address, 0, sizeof(address));
+			dnet_cmd cmd;
+			memset(&cmd, 0, sizeof(cmd));
+			cmd.id = id;
+			cmd.status = err;
+			cmd.cmd = DNET_CMD_BULK_READ_NEW;
+			cmd.trace_id = sess.get_trace_id();
+			cmd.flags = DNET_FLAGS_REPLY | DNET_FLAGS_MORE | (sess.get_trace_bit() ? DNET_FLAGS_TRACE_BIT : 0);
+			auto result_data = std::make_shared<ioremap::elliptics::callback_result_data>(&address, &cmd);
+			result_data->error = create_error(err, "send_bulk_read: could not locate address & backend for requested key: %s",
+							  dnet_dump_id(&id));
+			ioremap::elliptics::callback_result_entry entry(result_data);
+			failed_handler.process(callback_cast<read_result_entry>(entry));
+		}
+	}
+
+	failed_handler.complete(error_info());
+
+	std::vector<async_read_result> results;
+	results.reserve(remotes_ids.size() + 1);
+	results.emplace_back(std::move(failed_result));
+
+	dnet_time deadline;
+	dnet_current_time(&deadline);
+	deadline.tsec += sess.get_timeout();
+
+	for (auto &pair : remotes_ids) {
+		const auto &address = pair.first;
+		auto &ids = pair.second;
+
+		const auto request = dnet_bulk_read_request{
+			std::move(ids),
+			sess.get_ioflags(),
+			read_flags,
+			deadline
+		};
+		const auto serialized_request = serialize(request);
+
+		transport_control control;
+		control.set_command(DNET_CMD_BULK_READ_NEW);
+		control.set_cflags(DNET_FLAGS_NEED_ACK | DNET_FLAGS_NOLOCK);
+		control.set_data(serialized_request.data(), serialized_request.size());
+
+		auto session = sess.clean_clone();
+		if (!(sess.get_cflags() & DNET_FLAGS_DIRECT_BACKEND)) {
```

Comment:
See `bulk_read_handler::start()` - it does `m_session.set_direct_id(addr, backend_id);`

If user specify direct address without specifying a backend, then `DNET_FLAGS_DIRECT_BACKEND` is unset, but `DNET_FLAGS_DIRECT` is set.